text/0800-ts-adoption-plan.md

```diff
@@ -539,36 +539,42 @@ For the classic DI system, we no longer require (or benefit from) string key-bas
 
-[^reintroduce]: We could conceivably reintroduce a registry of this form in the future if that limitation on decorators in TypeScript changes, so that `@service('session') declare session;` would have the right type without needing an explicit type annotation, but that would need to be well-motivated on the merits, and there are actually advantages to having explicit type imports for the static analyzability of the code. The combination of Embroider and other future design moves we might make for the design system, for example [this experiment by pzuraq](https://github.com/pzuraq/ember-totally-not-module-based-services), may make other directions more appealing than the registry approach.
+Future designs for services and dependency injection more generally will need to be written to account for the capabilities of TypeScript’s implementation of Stage 3 decorators as they grow and change over time.
+
+For Ember Data, there is considerably more ongoing need for registry-style APIs. `Store.findRecord` isn’t going anywhere any time soon, for example, and it *requires* some kind of registry to make `this.store.findRecord('user', 1)` correctly return a `User` model. Many other APIs within Ember Data similarly require registries for type safety. Thus, Ember Data will need its own dedicated design for handling those, and this RFC leaves those for the Ember Data team to address in a dedicated RFC.
 
 
 ### CLI Integration
 
 We will introduce a new `--typescript` (`-ts`) flag for the `ember new` and `ember addon` commands, allowing users to opt into TypeScript when generating a new project. Using this flag will:
 
-- Install Ember TypeScript integration tooling, today consisting simply of `ember-cli-typescript` and its generators.
 - Set the `isTypeScriptProject` option for `.ember-cli` (introduced in [RFC #0776][rfc-0776]) to `true`, so that blueprints are generated with TypeScript by default in the project.
 - Configure linting:
     - In the`.eslintrc.js` blueprint:
-        - use `@typescript-eslint/parser` instead of the Babel ESLint parser
-        - include `@typescript-eslint` in the `plugins` key
-        - include `plugin:@typescript-eslint/recommended` in the `extends` key
-    - Install the `@typescript-eslint` dependencies instead of of the Babel ESLint dependencies in `package.json`
+        - Use `@typescript-eslint/parser` instead of the Babel ESLint parser.
+        - Include `@typescript-eslint` in the `plugins` key.
+        - Include `plugin:@typescript-eslint/recommended` in the `extends` key.
+    - Install the `@typescript-eslint` dependencies instead of,  or in addition to (as appropriate), the Babel ESLint dependencies in `package.json`.
+- Configure [Glint][glint] for type checking and, for addons, emitting type declarations during build.
```

Comment:
> for addons, emitting type declarations during build

Any idea or prescriptions how do that exactly? With e-c-ts we have dedicated commands for this, which I believe put the `.d.ts` files at the root level and know how to properly clean this up again as part of `postpack`. Without it, will we just put a long `tsc` command into `prepack`, mostly replicating [what the command does](https://github.com/typed-ember/ember-cli-typescript/blob/master/ts/lib/commands/precompile.ts#L28-L36)? But then how do we clean this up? Or should we put all declarations into a separate folder (we could use `dist` if it wasn't used already for the Ember CLI build, so maybe `declarations`?) and have `exports` and `typesVersions` configs in `package.json` point to it, similar to what v2 addons already do?

Reply:
1. I appreciate that you just solved this while I was buried in other things. 😁

2. I think the approach you landed on over at ember-cli/ember-cli#10283 is a good one. For other folks who haven't read that (and for posterity): the blueprint [emits types into `declarations`][decl] and [uses `typesVersions` to make resolution work correctly for consumers][tv]. (It also removes that after publishing.) This is probably what we would have done in the first place with `ember-cli-typescript` if we had had `typesVersions` available all those years ago.

[decl]: https://github.com/ember-cli/ember-cli/pull/10283/files#diff-30de8a6a8a92275125cfd88e3d43feaed99481211446fe701ce5ec00ff4642edR4
[tv]: https://github.com/ember-cli/ember-cli/pull/10283/files#diff-0ee18a1bec40c2bce48535d75189607df32ab18631183de0297b3c3755d352aaR92-R99